File: src/javascript/datastructures/linkedlist/class/linkedList.js
class LinkedList {
    constructor(value) {
        const newNode = new Node(value);
        this.head = newNode;
        this.tail = this.head;
        this.length = 1;
    }

    push(value) {
        const newNode = new Node(value);
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }

    pop() {
        if (this.length === 0) return undefined;
        let temp = this.head;
        let prev = this.head;
        while(temp.next) {
            prev = temp;
            temp = temp.next;
        }
        this.tail = prev;
        this.tail.next = null;
        this.length--;
        if(this.length === 0) {
            this.head = null;
            this.tail = null;
        }
        return temp;
    }

    unshift(value) {
        const newNode = new Node(value);
        if(!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
        this.length++;
        return this;
    }

    shift() {
        if(!this.head) return undefined;
        let temp = this.head;
        this.head = this.head.next;
        temp.next = null;
        this.length--;
        if(this.length === 0) {
            this.tail = null;
        }
        return temp;
    }

    get(index) {
        if(index < 0 || index >= this.length) return undefined;
        let temp = this.head;
        for(let i = 0; i < index; i++) {
            temp = temp.next;
        }
        return temp;
    }

    set(index, value) {
        let temp = this.get(index);
        if(temp) {
            temp.value = value;
            return true;
        }
        return false;
    }

    insert(index, value) {
        if(index < 0 || index > this.length) return false;
        if(index === 0) return this.unshift(value);
        if(index === this.length) return this.push(value);

        const newNode = new Node(value);
        const temp = this.get(index-1);
        newNode.next = temp.next;
        temp.next = newNode;
        this.length++;
        return true;
    }

    remove(index) {
        if(index < 0 || index > this.length) return undefined;
        if(index === 0) return this.shift();
        if(index === this.length-1) return this.pop();

        const prev = this.get(index-1);
        const temp = prev.next;
        prev.next = temp.next;
        temp.next = null;
        this.length--;
        return temp;
    }

    reverse() {
        let temp = this.head;
        this.head = this.tail;
        this.tail = temp;
        let next = temp.next;
        let prev = null;
        for(let i = 0; i < this.length; i++) {
            next = temp.next;
            temp.next = prev;
            prev = temp;
            temp = next;
        }
        return this;
    }

}

const myLinkedList = new LinkedList(1);
myLinkedList.push(2);
myLinkedList.push(3);

myLinkedList.pop();
myLinkedList.pop();

myLinkedList.unshift(2);
myLinkedList.unshift(3);

myLinkedList.shift();

myLinkedList.get(1);
myLinkedList.get(0);

myLinkedList.set(0, 1);
myLinkedList.set(1, 3);

myLinkedList.insert(1, 2);
myLinkedList.insert(3, 4);

myLinkedList.remove(3);
myLinkedList.remove(2);

myLinkedList.reverse();
myLinkedList.reverse();

File: src/java1/datastructures/linkedlist/LinkedList.java
package java1.datastructures.linkedlist;

public class LinkedList {
    private Node head;
    private Node tail;
    private int length;

    LinkedList() {
        head = null;
        tail = null;
        length = 0;
    }

    public void append(int value) {
        Node newNode = new Node(value);
        if(this.length == 0) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            tail = newNode;
        }
        length++;
    }

    public Node removeLast() {
        if(length == 0) return null;
        Node current = head;
        Node newTail = head;
        while(current.next != null) {
            newTail = current;
            current = current.next;
        }
        tail = newTail;
        tail.next = null;
        length--;
        if(length == 0) {
            head = null;
            tail = null;
        }
        return current;
    }

    public void prepend(int value) {
        Node newNode = new Node(value);
        if(length == 0) {
            head = newNode;
            tail = newNode;
        } else {
            newNode.next = head;
            head = newNode;
        }
        this.length++;
    }

    public Node removeFirst() {
        if(length == 0) return null;
        Node currentHead = head;
        head = head.next;
        currentHead.next = null;
        this.length--;
        if(length == 0) {
            tail = null;
        }
        return currentHead;
    }

    public Node get(int index) {
        if(index < 0 || index >= length) return null;
        Node current = head;
        for(int i=1; i<= index; i++) {
            current = current.next;
        }
        return current;
    }

    public boolean set(int index, int value) {
        Node foundNode = get(index); 
        if(foundNode != null) {
            foundNode.value = value;
            return true;
        }
        return false;
    }

    public boolean insert(int index, int value) {
        if(index < 0 || index > length) return false;

        if(index == 0) {
            prepend(value);
            return true;
        }
        if(index == length) {
            append(value);
            return true;
        }
        Node newNode = new Node(value);
        Node prevNode = get(index-1);
        newNode.next = prevNode.next;
        prevNode.next = newNode;
        length++;
        return true;
    }

    public Node remove(int index) {
        if(index < 0 || index >= length) return null;
        if(index == 0) return removeFirst();
        if(index == length -1) return removeLast();

        Node prevNode = get(index-1);
        Node removedNode = prevNode.next;
        prevNode.next = removedNode.next;
        removedNode.next = null;
        length--;
        return removedNode;
    }

    //TC: O(n) SC: O(1)
    public void reverse() {
        if(length == 0) return;
        Node temp = head;
        head = tail;
        tail = temp;
        Node next = temp.next;
        Node prev = null;
        
        for(int i=0; i< length; i++){
            next = temp.next;
            temp.next = prev;
            prev = temp;
            temp = next;
        }
    }

    //Recursion: TC: O(n) SC: O(n)
    public Node reverseLL(Node head) {
        if(head == null || head.next == null) {
            return head;
        }
        Node rest = reverseLL(head.next);

        head.next.next = head;
        head.next = null;
        return rest;
    }

    public void printList() {
        Node temp = head;
        while(temp != null) {
            System.out.println(temp.value);
            temp = temp.next;
        }
    }

    public Node getHead() {
        return head;
    }
}
